File: leetcode/cpp/154. 寻找旋转排序数组中的最小值 II.cpp
class Solution {
public:
	// ===== 之前的解法 =====
    int min(vector<int>& nums, int l, int h) {
        int tmp = nums[l];
        for (int i = l; i <= h; i++) {
            if (tmp > nums[i]) {
                tmp = nums[i];
            }
        }
        return tmp;
    }

    int findMin(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }

        int l = 0;
        int h = nums.size() - 1;
        while (l <= h) {
            int m = (l + h) / 2;
            if (nums[l] == nums[m] && nums[m] == nums[h]) {
                return min(nums, l, h);
            } else if (nums[m] <= nums[h]) {
                h = m;
            } else {
                l = m + 1;
            }
        }
        return nums[l];
    }

	// ===== 后续的解法 =====
    int findMin(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }

        int l = 0, h = nums.size() - 1;
        while (l < h) {
            int m = l + (h - l) / 2;
            if (nums[l] == nums[m] && nums[m] == nums[h]) {
                return *min_element(nums.begin() + l, nums.begin() + h);
            } else if (nums[m] <= nums[h]) {
                h = m;
            } else {
                l = m + 1;
            }
        }
        return nums[l];
    }
};

File: leetcode/python/154. 寻找旋转排序数组中的最小值 II.py
#!/bin/bash python
"""
154. 寻找旋转排序数组中的最小值 II

思路：
增加了重复元素
"""

class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        区别在于存在重复数组
        """
        if nums == []:
            return 0
        
        l, h = 0, len(nums) - 1
        while l < h:
            m = (l + h) // 2
            if nums[l] == nums[m] == nums[h]: # 此时无法进行区分
                return min(nums[l:h+1])
            elif nums[m] <= nums[h]:
                h = m
            else:
                l = m + 1
        return nums[l]